packages/superset-ui-core/src/SupersetClient.js

```diff
@@ -0,0 +1,155 @@
+import callApi from './callApi';
+
+const AUTH_QUEUE_POLL_MS = 30;
+
+class SupersetClient {
+  constructor(config) {
+    const {
+      protocol = 'http',
+      host = '',
+      headers = {},
+      mode = 'same-origin',
+      timeout,
+      credentials,
+    } = config;
+
+    this.headers = headers;
+    this.host = host;
+    this.mode = mode;
+    this.timeout = timeout;
+    this.protocol = protocol;
+    this.credentials = credentials;
+    this.csrfToken = null;
+    this.didAuthSuccessfully = false;
+    this.csrfPromise = null;
+
+    // this.getUnauthorizedError = this.getUnauthorizedError.bind(this);
+    // this.waitForCSRF = this.waitForCSRF.bind(this);
+  }
+
+  isAuthenticated() {
+    return this.didAuthSuccessfully;
+  }
+
+  init() {
+    return this.getCSRFToken();
+  }
+
+  getCSRFToken() {
+    // If we can request this resource successfully, it means that the user has
+    // authenticated. If not we throw an error prompting to authenticate.
+    return (this.csrfPromise = callApi({
+      credentials: this.credentials,
+      headers: {
+        ...this.headers,
+      },
+      method: 'GET',
+      mode: this.mode,
+      timeout: this.timeout,
+      url: this.getUrl({ endpoint: 'superset/csrf_token/', host: this.host }),
+    }).then(
+      response => {
+        if (response.json) {
+          this.csrfToken = response.json.csrf_token;
+          this.headers = { ...this.headers, 'X-CSRFToken': this.csrfToken };
+          this.didAuthSuccessfully = !!this.csrfToken;
+        }
+
+        if (!this.csrfToken) {
+          return Promise.reject({ error: 'Failed to fetch CSRF token' });
+        }
+
+        return response;
+      },
+      error => Promise.reject(error),
+    ));
+  }
+
+  getUrl({ host = '', endpoint = '' }) {
+    const cleanHost = host.slice(-1) === '/' ? host.slice(0, -1) : host; // no backslash
+
+    return `${this.protocol}://${cleanHost}/${endpoint[0] === '/' ? endpoint.slice(1) : endpoint}`;
+  }
+
+  ensureAuth() {
+    return (
+      this.csrfPromise ||
+      Promise.reject({
+        error: `SupersetClient has no CSRF token, ensure it is initialized or
+        try logging into the Superset instance at ${this.getUrl('/login')}`,
+      })
+    );
+  }
+
+  get({ host, url, endpoint, mode, credentials, headers, body, timeout, signal }) {
+    return this.ensureAuth().then(() =>
```

Comment:
What do you think about attaching an `abort` function to the returned promise?

```js
const promise = this.ensureAuth().then(...);
promise.abort = signal && () => signal.abort();
return promise;
```

So can use like this. 

```js
const myPromise = SupersetClient.get(...);
if(wannaCancel) {
  myPromise.abort();
}
```

The concern is the function `abort()` will only be available to the original Promise returned from `get` and not on the chained `then()`, which may be not obvious to developer.

Reply:
hmm, that's an interesting thought, the abort controller api is a little verbose. I do think it would be non-ideal to have the `abort` just attached to the initial promise. 

maybe we can see how it goes with `signal` and we could develop a nicer / more complex api around aborting in the future? there were only a couple of places in Superset that use this currently.

Reply:
in practice you either have to pass the abort controller around, or the promise, and so I'm not sure there's much advantage of one over the other except that the former is full-fledged API spec not just our library API.

Reply:
sgtm. We can try that later. I think this first version looks good to me.